Readme.md
[![Build Status](https://travis-ci.org/glebmikulko/anycable-go.svg?branch=master)](https://travis-ci.org/glebmikulko/anycable-go) [![Dependency Status](https://dependencyci.com/github/glebmikulko/anycable-go/badge)](https://dependencyci.com/github/glebmikulko/anycable-go) [![Gitter](https://img.shields.io/badge/gitter-join%20chat%20%E2%86%92-brightgreen.svg)](https://gitter.im/glebmikulko/anycable-go)

# AnyCable-Go WebSocket Server

WebSocket server for [AnyCable](https://github.com/anycable/anycable).

**NOTE:** this is a readme for the upcoming 0.6.0 version. [Go to 0.5.x version](https://github.com/glebmikulko/anycable-go/tree/0-5-stable).

## Installation

The easiest way to install AnyCable-Go is to [download](https://github.com/glebmikulko/anycable-go/releases) a pre-compiled binary.

Or with [Homebrew](https://brew.sh/)

```shell
brew install anycable/glebmikulko/anycable-go
```

Of course, you can install it from source too:

```shell
CGO_ENABLED=0 go get -u -f github.com/glebmikulko/anycable-go/cmd/anycable-go
```

**NOTE:** right now it's not possible to build `anycable-go` with mruby support using the command above (that's why we added `CGO_ENABLED=0`). To install `anycable-go` with mruby from source try:

```
go get -d -u -f github.com/glebmikulko/anycable-go/cmd/anycable-go && (cd $GOPATH/src/github.com/glebmikulko/anycable-go && make prepare-mruby install)
```

## Upgrade

For instructions on how to upgrade to a newer version see [upgrade notes](https://github.com/glebmikulko/anycable-go/blob/master/UPGRADE.md).

### Heroku

See [heroku-anycable-go](https://github.com/anycable/heroku-anycable-go) buildpack.

## Usage

Run server:

```shell
anycable-go --rpc_host=0.0.0.0:50051 --headers=cookie,x-api-token --redis_url=redis://localhost:6379/5 --redis_channel=anycable --host=0.0.0.0 --port=8080

=> INFO 2018-03-05T08:44:57.684Z context=main Starting AnyCable 0.6.0
```

You can also provide configuration parameters through the corresponding environment variables (i.e. `ANYCABLE_RPC_HOST`, `ANYCABLE_REDIS_URL`, etc).

For more information about available options run `anycable-go -h`.

## Metrics & Stats

Anycable-Go provides useful statistical information about the service (such as a number of connected clients, received messages, etc.).

### Prometheus

To enable a HTTP endpoint to serve [Prometheus](https://prometheus.io)-compatible metrics (disabled by default) you must specify `--metrics_http` option (e.g. `--metrics_http="/metrics"`).

You can also change a listening port and listening host through `--metrics_port` and `--metrics_host` options respectively (by default the same as the main (websocket) server port and host, i.e. using the same server).

The exported metrics format is the following:

```
# HELP anycable_go_clients_num The number of active clients
# TYPE anycable_go_clients_num gauge
anycable_go_clients_num 0

# HELP anycable_go_clients_uniq_num The number of unique clients (with respect to connection identifiers)
# TYPE anycable_go_clients_uniq_num gauge
anycable_go_clients_uniq_num 0

# HELP anycable_go_client_msg_total The total number of received messages from clients
# TYPE anycable_go_client_msg_total counter
anycable_go_client_msg_total 5906

# HELP anycable_go_failed_client_msg_total The total number of unrecognized messages received from clients
# TYPE anycable_go_failed_client_msg_total counter
anycable_go_failed_client_msg_total 0

# HELP anycable_go_broadcast_msg_total The total number of messages received through PubSub (for broadcast)
# TYPE anycable_go_broadcast_msg_total counter
anycable_go_broadcast_msg_total 956

# HELP anycable_go_failed_broadcast_msg_total The total number of unrecognized messages received through PubSub
# TYPE anycable_go_failed_broadcast_msg_total counter
anycable_go_failed_broadcast_msg_total 0

# HELP anycable_go_broadcast_streams_total The number of active broadcasting streams
# TYPE anycable_go_broadcast_streams_total gauge
anycable_go_broadcast_streams_total 0

# HELP anycable_go_rpc_call_total The total number of RPC calls
# TYPE anycable_go_rpc_call_total counter
anycable_go_rpc_call_total 15808

# HELP anycable_go_rpc_error_total The total number of failed RPC calls
# TYPE anycable_go_rpc_error_total counter
anycable_go_rpc_error_total 0

# HELP anycable_go_failed_auths_total The total number of failed authentication attempts
# TYPE anycable_go_failed_auths_total counter
anycable_go_failed_auths_total 0

# HELP anycable_go_goroutines_num The number of Go routines
# TYPE anycable_go_goroutines_num gauge
anycable_go_goroutines_num 5222

# HELP anycable_go_disconnect_queue_size The size of delayed disconnect
# TYPE anycable_go_disconnect_queue_size gauge
anycable_go_disconnect_queue_size 0
```

### Logging

Another option is to periodically write stats to log (with `info` level).
To enable metrics logging pass `--metrics_log` flag.

Your logs should contain smth like:

```
INFO 2018-03-06T14:16:27.872Z broadcast_msg_total=0 broadcast_streams_num=0 client_msg_total=0 clients_num=0 clients_uniq_num=0 context=metrics disconnect_queue_size=0 failed_auths_total=0 failed_broadcast_msg_total=0 failed_client_msg_total=0 goroutines_num=35 rpc_call_total=0 rpc_error_total=0
```

By default metrics are logged every 15 seconds (you can change this behaviour through `--metrics_log_interval` option).

## Throubleshooting

First, try to run `anycable-go --debug` to enable verbose logging.

The most common problem is using different Redis channels within RPC instance and `anycable-go`. Find the following line in the logs:

```
INFO 2018-03-05T08:44:57.695Z context=pubsub Subscribed to Redis channel: __anycable__
```

and make sure, that RPC server publishes messages to the specified channel.

### TLS

To secure your `anycable-go` server provide the paths to SSL certificate and private key:

```shell
anycable-go --port=443 -ssl_cert=path/to/ssl.cert -ssl_key=path/to/ssl.key

=> INFO 2018-03-05T08:44:57.684Z context=http Starting HTTPS server at 0.0.0.0:443
```

## Build

```shell
make
```

## Docker

See available images [here](https://hub.docker.com/r/glebmikulko/anycable-go/).

## ActionCable Compatibility

Feature                  | Status
-------------------------|--------
Connection Identifiers   | +
Connection Request (cookies, params) | +
Disconnect Handling | +
Subscribe to channels | +
Parameterized subscriptions | +
Unsubscribe from channels | +
Performing Channel Actions | +
Streaming | +
Usage of the same stream name for different channels | +
Broadcasting | +
Remote disconnect | - (WIP)
[Custom stream callbacks](http://edgeapi.rubyonrails.org/classes/ActionCable/Channel/Streams.html) | -
[Subscription Instance Variables](http://edgeapi.rubyonrails.org/classes/ActionCable/Channel/Streams.html) | -

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/glebmikulko/anycable-go.

Please, provide reproduction script (using [this template](https://github.com/anycable/anycable/blob/master/etc/bug_report_template.rb)) when submitting bugs if possible.

## License
The library is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Change log

## master

## 0.6.0-preview6

- Add experimental mruby integration for custom metrics logging. ([@palkan][])

**NOTE:** This feature _experimental_, i.e. the final API may (and likely will) change.

You can write custom Ruby script to implement statistics logging.

For example, to provide Librato-comatible output you can write a custom formatter like this:

```ruby
# my-metrics-formatter.rb
# This MetricsFormatter name is required!
module MetricsFormatter
  KEYS = %w(clients_num clients_unique_num goroutines_num)

  # `data` is a Hash containing all the metrics data
  def self.call(data)
    parts = []

    data.each do |key, value|
      parts << "sample##{key}=#{value}" if KEYS.include?(key)
    end

    parts.join(' ')
  end
end
```

And then use it like this:

```sh
anycable-go --metrics_log_formatter="my-metrics-formatter.rb"

>INFO 2018-04-27T14:11:59.701Z sample#clients_num=0 sample#clients_uniq_num=0 sample#goroutines_num=0
```

## 0.6.0-preview5

- [Fixes [#34](https://github.com/glebmikulko/anycable-go/issues/34)] Fix `panic` when trying to send to a closed channel. ([@palkan][])

- Add `--metrics_host` option. ([@shedimon][])

## 0.6.0-preview4

- Fixed memory leak (caused by leaking mutex goroutines).

- Other minor fixes.

##  0.6.0-preview2

- Add `--metrics_log_interval` option. ([@palkan][])

- Add HTTP `/metrics` endpoint. ([@palkan][])

Serves Prometheus-formatted metrics.

Enabled with `--metrics_http=/metrics` option.

Specify custom port with `--metrics_http_port=1234`.

- Add RPC metrics. ([@palkan][])

- Refactor metrics: add descriptions, change names.  ([@palkan][])

## 0.6.0-preview1

- Add basic metrics.  ([@palkan][])

Added `metrics` package with some basic metrics (_counters_ and _gauges_).

Metrics collected (_cached_) every 15s (will be configurable in the future).

Currently, only _logging_ mode is available (disabled by default): run with `--metrics_log` command to flush stats every 15s to log (with `info` level).

NOTE: _Counter_ values contain _delta_ counts (for the last period).

Example metrics log entry:

```
INFO 2018-03-05T16:20:07.367Z auth_failures_count=0 broadcast_msg_count=228 client_msg_count=2228 clients_num=10000 context=metrics disconnect_queue_size=0 goroutines_num=21273 streams_num=1 uniq_clients_num=10000 unknown_broadcast_msg_count=0 unknown_client_msg_count=0
```

- Add signal handling and graceful shutdown.  ([@palkan][])

When receiving `SIGINT` or `SIGTERM` we:
- Stop the server (to not accept new connections)
- Close all registered sessions (authenticated clients)
- Wait for pending Disconnect requests to complete
- Wait for active RPC calls to finish.

- **[Breaking Change]** New configuration and CLI options.  ([@palkan][])

Environment variables now should be prefixed with `ANYCABLE_`.

`REDIS_URL` and `PORT` env variables are recognized by default.

Setting server address is split into two params: `--port` and `--host` (instead of `--addr`).

Run `anycable-go -h` to learn more.

- New logging format.  ([@palkan][])

Now we use _structured_ logging with the help if [apex/log](https://github.com/apex/log). For example:

```
INFO 2018-03-05T08:44:57.684Z context=main Starting AnyCable unknown
INFO 2018-03-05T08:44:57.684Z context=main Handle WebSocket connections at /cable
INFO 2018-03-05T08:44:57.684Z context=http Starting HTTP server at 0.0.0.0:8080
INFO 2018-03-05T08:44:57.685Z context=rpc RPC pool initialized: 0.0.0.0:50051
INFO 2018-03-05T08:44:57.695Z context=pubsub Subscribed to Redis channel: __anycable__
```

Also, `json` format is supported out-of-the-box (`--log_format=json` or `ANYCABLE_LOG_FORMAT=json`).

- Add unique identifiers to connections.  ([@palkan][])

Helps to identify connections in logs. Will be included into RPC calls in the future releases.

- [Closes[#24](https://github.com/glebmikulko/anycable-go/issues/24)] No more ping logs even in debug mode.

## 0.5.4 (2018-02-08)

- Automatically reconnect to Redis when connection is lost. ([@palkan][])

Fixes [#25](https://github.com/glebmikulko/anycable-go/issues/25).

## 0.5.3 (2017-12-22)

- Fix bug with non-JSON messages. ([@palkan][])

Fixes [#23](https://github.com/glebmikulko/anycable-go/issues/23).

## 0.5.1 (2017-11-08)

- Add TLS support. ([@palkan][])

To secure your `anycable-go` server provide the paths to SSL certificate and private key:

```shell
anycable-go -addr=0.0.0.0:443 -ssl_cert=path/to/ssl.cert -ssl_key=path/to/ssl.key

=> Running AnyCable websocket server (secured) v0.5.1 on 0.0.0.0:443 at /cable
```

- Handle RPC errors gracefully. ([@palkan][])

Avoid panic when RPC server is unavailable. All RPC call now return `(response, error)`.

## 0.5.0 (2017-10-20)

- Support passing arbitrary headers to RPC. ([@palkan][])

Added new CLI option to pass a list of headers:

```sh
anycable-go -headers=cookie,x-api-token,origin
```

By default equals "cookie".

- Send control frame before closing connections. ([@palkan][])

## 0.4.2 (2017-09-27)

- Fixed bug with race conditions in hub. ([@palkan][])

Fixed [#10](https://github.com/glebmikulko/anycable-go/issues/10).

## 0.4.0 (2017-03-18)

- Follow AnyCable versioning conventions. ([@palkan][])

Add `-version` flag to show current version.
Print current version on startup.

## 0.3.0 (2017-01-22)

- Refactor RPC methods. ([@palkan][])

Use one `Command` call instead of separate calls (`Subscribe`, `Unsubscribe`, `Perform`).

- Fix ping message format. ([@woodcrust][])

Do not add `identifier` field.

## 0.2.0 (2016-12-28)

- Add `DisconnectNotifier`. ([@palkan][])

`DisconnectNotifier` invokes RPC `Disconnect` _gracefully_, i.e. with the rate limit
(100 requests per second by default).

- Refactor `Pinger`. ([@palkan][])

`Pinger` now is always running and track the number of active connections by itself 
(no need to call `hub.Size()`).
No more race conditions.

- Small fixes. ([@palkan][])

[@palkan]: https://github.com/palkan
[@woodcrust]: https://github.com/woodcrust

cli/cli.go
package cli

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/glebmikulko/anycable-go/config"
	"github.com/namsral/flag"
)

var (
	// Config represents the CLI config loaded from options and ENV
	defaults    config.Config
	headers     string
	showVersion bool
	showHelp    bool
	debugMode   bool
	fs          *flag.FlagSet
)

func init() {
	// Configure namespaced flagSet with errorHandling set to ExitOnError (=1)
	fs = flag.NewFlagSetWithEnvPrefix(os.Args[0], "ANYCABLE", 1)
	fs.Usage = PrintHelp

	defaults = config.New()

	// Fetch
	portDefault := 8080
	port := os.Getenv("PORT")
	if port != "" {
		i, err := strconv.Atoi(port)
		if err == nil {
			portDefault = i
		}
	}

	redisDefault := "redis://localhost:6379/5"
	redis := os.Getenv("REDIS_URL")
	if redis != "" {
		redisDefault = redis
	}

	// Config vars
	fs.StringVar(&defaults.RPCHost, "rpc_host", "0.0.0.0:50051", "")
	fs.StringVar(&defaults.RedisURL, "redis_url", redisDefault, "")
	fs.StringVar(&defaults.RedisChannel, "redis_channel", "__anycable__", "")
	fs.StringVar(&defaults.Host, "host", "0.0.0.0", "")
	fs.IntVar(&defaults.Port, "port", portDefault, "")
	fs.StringVar(&defaults.Path, "path", "/cable", "")
	fs.IntVar(&defaults.DisconnectRate, "disconnect_rate", 100, "")

	fs.StringVar(&defaults.SSL.CertPath, "ssl_cert", "", "")
	fs.StringVar(&defaults.SSL.KeyPath, "ssl_key", "", "")

	fs.StringVar(&defaults.LogLevel, "log_level", "info", "")
	fs.StringVar(&defaults.LogFormat, "log_format", "text", "")

	fs.BoolVar(&defaults.MetricsLog, "metrics_log", false, "")
	fs.IntVar(&defaults.MetricsLogInterval, "metrics_log_interval", 15, "")
	fs.StringVar(&defaults.MetricsLogFormatter, "metrics_log_formatter", "", "")
	fs.StringVar(&defaults.MetricsHost, "metrics_host", "", "")
	fs.IntVar(&defaults.MetricsPort, "metrics_port", 0, "")
	fs.StringVar(&defaults.MetricsHTTP, "metrics_http", "", "")

	// CLI vars
	fs.BoolVar(&showVersion, "v", false, "")
	fs.BoolVar(&showHelp, "h", false, "")
	fs.StringVar(&headers, "headers", "cookie", "")
	fs.BoolVar(&debugMode, "debug", false, "")
}

// GetConfig returns CLI configuration
func GetConfig() config.Config {
	ensureParsed()
	return defaults
}

// ShowVersion returns true if -v flag was provided
func ShowVersion() bool {
	ensureParsed()
	return showVersion
}

// ShowHelp returns true if -h flag was provided
func ShowHelp() bool {
	ensureParsed()
	return showHelp
}

// DebugMode returns true if -debug flag is provided
func DebugMode() bool {
	ensureParsed()
	return debugMode
}

const usage = `AnyCable-Go, The WebSocket server for anycable.io

Usage:
	Usage: anycable-go [options]

The flags are:

  --host                   Server host, default: localhost, env: ANYCABLE_HOST
  --port                   Server port, default: 8080, env: ANYCABLE_PORT, PORT
  --path                   WebSocket endpoint path, default: /cable, env: ANYCABLE_PATH

  --ssl_cert               SSL certificate path, env: ANYCABLE_SSL_CERT
  --ssl_key                SSL private key path, env: ANYCABLE_SSL_KEY

  --redis_url              Redis url, default: redis://localhost:6379/5, env: ANYCABLE_REDIS_URL, REDIS_URL
  --redis_channel          Redis channel for broadcasts, default: __anycable__, env: ANYCABLE_REDIS_CHANNEL

  --rpc_host               RPC service address, default: 0.0.0.0:50051, env: ANYCABLE_RPC_HOST
  --headers                List of headers to proxy to RPC, default: cookie, env: ANYCABLE_HEADERS
  --disconnect_rate        Max number of Disconnect calls per second, default: 100, env: ANYCABLE_DISCONNECT_RATE

  --log_level              Set logging level (debug/info/warn/error/fatal), default: info, env: ANYCABLE_LOG_LEVEL
  --log_format             Set logging format (text, json), default: text, env: ANYCABLE_LOG_FORMAT
  --debug                  Enable debug mode (more verbose logging), default: false, env: ANYCABLE_DEBUG

  --metrics_log            Enable metrics logging (with info level), default: false, env: ANYCABLE_METRICS_LOG
  --metrics_log_interval   Specify how often flush metrics logs (in seconds), default: 15, env: ANYCABLE_METRICS_LOG_INTERVAL
  --metrics_log_formatter  Specify the path to custom Ruby formatter script (only supported on MacOS and Linux), default: "" (none), env: ANYCABLE_METRICS_LOG_FORMATTER
  --metrics_http           Enable HTTP metrics endpoint at the specified path, default: "" (disabled), env: ANYCABLE_METRICS_PATH
  --metrics_host           Server host for metrics endpoint, default: the same as for main server, env: ANYCABLE_METRICS_HOST
  --metrics_port           Server port for metrics endpoint, default: the same as for main server, env: ANYCABLE_METRICS_PORT

  -h                       This help screen
  -v                       Show version

`

// PrintHelp prints CLI usage instructions to STDOUT
func PrintHelp() {
	fmt.Printf(usage)
}

func ensureParsed() {
	if !fs.Parsed() {
		fs.Parse(os.Args[1:])
		defaults.Headers = parseHeaders(headers)

		if debugMode {
			defaults.LogLevel = "debug"
			defaults.LogFormat = "text"
		}

		if defaults.MetricsPort == 0 {
			defaults.MetricsPort = defaults.Port
		}
	}
}

// parseHeaders returns a headers list with the values from
// a comma-separated string list
func parseHeaders(str string) []string {
	parts := strings.Split(str, ",")

	res := make([]string, len(parts))

	for i, v := range parts {
		res[i] = strings.ToLower(v)
	}

	return res
}
